Add: readme file tutorial to packaging 101 series (#176)
pyOpenSci/python-package-guide: tutorials/add-readme.md

# Add a README file to your Python package

In the previous lessons you learned:

1. What a Python package is

> 2. How to make your code installable and

Reviewer: ```suggestion
2. How to make your code installable
```
if you wanted to keep "and," then I'd recommend putting a comma after "1. What a Python package is"

Author: I think everytime i create a markdown list i use slightly different grammar 😆

3. How to publish your package to (test) PyPI

:::{admonition} Learning objectives

> In this lesson you will learn:

Reviewer: this is a small nit, but consider adding 1. What a **README.md** file is, since it's what we jump right into lesson-wise

Author: @kierisi i'd suggest this is not at all a nit. it's actually a big element of the lesson that i missed when highlighting in the Learning Objectives! Great catch! fixed locally

1. How to add a **README.md** file to your package.
2. What the core elements of a **README.md** file are.
:::

## What is a README file?

The `README.md` file is a markdown file located at the root of your project directory that helps
a user understand:

- You package's name and what it does

> - The current development "state" of the package (through badges)

Reviewer: it could be helpful to link to badges here for newer github users (I'll also add badges to the glossary list!)

Author: @kierisi this is great feedback. what do you think would be helpful for a newer github user? Maybe some example badges and explanation of what badges are and do in a breakout? i can look around for links as well.

Author: Does this address your comment @kierisi ? i created a small breakout on creating badges with some examples of types of badges and what they communicate. 

<img width="871" alt="Screenshot 2024-02-07 at 3 33 00 PM" src="https://github.com/pyOpenSci/python-package-guide/assets/7649194/01d95557-ae68-4212-a4e0-e5824d1cb231">

Reviewer: @lwasser this is perfect!

Author: ok perfect! i feel like i saw a badging resource somewhere recently too that maybe @NickleDave directed me to. If i can find that i will add a link about it as well.

> - The current development "state" of the package (through badges)

Reviewer 2: I personally think the biggest job of a README is to describe what the project does, maybe with a bit of a sales pitch, but at least a sort of problem statement. This is actually one of the required items for pyOS reviews so surprised it's not here.

Author: hey @ucodery this is great feedback. Step three below the badges is to add a description of the package and what it does. Are you looking for more detail here in terms of the "pitch" and need? OR are you looking for something more specific higher up?  i added two screenshots below with content that i think addresses what you are looking for. but i suspect i'm missing something and may just need a bit of direction regarding what you think would make things better! many thanks. this feedback is super useful i just need guidance in implementing a fix!

<img width="742" alt="Screenshot 2024-02-13 at 4 54 49 PM" src="https://github.com/pyOpenSci/python-package-guide/assets/7649194/17a7da26-8d43-44e3-923c-d1399ffee292">

 
<img width="747" alt="Screenshot 2024-02-13 at 4 53 47 PM" src="https://github.com/pyOpenSci/python-package-guide/assets/7649194/80f082f4-a098-4642-97d6-daa9720a2c66">

Author: it seems like folks in slack generally agree with you. so i'll pull the license part from this lesson! it does make sense to make use of the pypi metadata and the github interface for these things.

Reviewer 2: Hmmm, I think I may not be reading "what it does" in the way that you mean it.
I read that as just a prose description of what is inside. Maybe even just some docstrings. While I believe a better statement in the readme is why you should use it - the elevator pitch.

Just as a reminder, this is a required item for PyOS review:
> **A statement of need** clearly stating problems the software is designed to solve and its target audience in README.

Which I think is great. To me, a "statement of need" reads differently than "what it does"

BUT this is really getting into tone and preferred writing style, and probably far off track for this PR. I don't think any changes are needed at this time. If I actually come up with a better tutorial phrasing, I can always propose new changes after this goes live :)

Author: ok no problem! i hear you. 

i've added this to the bulleted list!

- You package's name 
- What the package does. Your README file should clearly state the problem(s) that your software is designed to solve and its target audience.
- The current development "state" of the package (through badges)
- How to get started with using your package.
- How to contribute to your package
- How to cite your package
-

Author: and then yes you can also submit another pr to adjust language :) i'll try to merge this today given it's been open for a while!! 🚀

- How to get started with using your package.
- How to contribute to your package
- How to cite your package

> Your **README.md** file is important as it is often the first thing that someone sees before they install your package. The README file also will be used to populate your PyPI landing page.

Reviewer: is this referring to PyPI or the test PyPI page?

Reviewer 2: It populates both! (if you publish to both)

Author: I've added this for additional clarity:

> The README file also will be used to populate your PyPI landing page. This is true for both test.pypi.org and pypi.org.

Author: i think it's a great idea to reiterate that it works for both especially if our reader is newer 🚀 !

Note that there is no required format for README files. This page simply outlines sections that we suggest you have in your README file.

## Create a README.md file for your package

It's time to add a `README.md` file to your project directory.

### Step 0. Create a README file
To get started, if you don't already have a README.md file in your project directory, create one.

:::{tip}
If you created your project directory from

* a GitHub repository online
* using `hatch init`

Then you may already have a README.MD file in your project directory.
:::

<!-- If they use hatch init in the very first lesson -
the readme will already be there-->

> ### Step 1. Add package badges

Reviewer 2: Wrong title? This should be Add Name or something? Maybe call out that it should probably be a h1 title; though not sure what familiarity with markdown we are assuming.

Author: Fixed! wrong title was there for sure. and i added a note about the markdown headers as well. 

<img width="747" alt="Screenshot 2024-02-13 at 4 58 55 PM" src="https://github.com/pyOpenSci/python-package-guide/assets/7649194/62311f08-d185-4905-b5ee-d347a6fceb44">

At the top of the `README.md` file, add the name of your package.

### Step 2 - add badges to the top of your README file

It's common for maintainers to add badges to the top of their README files. Badges allow you and your package users to track things like

> * Broken documentation and test builds

Reviewer: ```suggestion
* Broken documentation and test builds.
```
added for consistency with other list items

* Versions of your package that are on PyPI and Conda.
* Whether your package has been reviewed and vetted by an organization such as pyOpenSci and/or JOSS.

If you have already published your package to pypi.org you can use [shields.io to create a package version badge](https://shields.io/badges/py-pi-version). This badge will dynamically update as you release new versions of your package to PyPI.

> If not you can leave the top empty for now and add badges to your README at a later point as they make sense for your package.

Reviewer: ```suggestion
If not, you can leave the top empty for now and add badges to your README at a later point as they make sense for your package.
```

### Step 3 - Add a description of what your package does

Below the badges (if you have them), add a section of text
that provides an easy-to-understand overview of what your
package does.

* Keep this section short.
* Try to avoid jargon.
* Define technical terms that you use to make the description accessible to more people.

Remember that the more people understand what your package does, the more people will use it.

### Step 4 - Add package installation instructions

Next, add instructions that tell users how to install your package.

For example, can they use pip to install your package?
`pip install packagename`

or conda?

`conda install -c conda-forge packagename`.

If you haven't yet published your package to pypi.org then
you can skip this section and come back and add these
instructions later.

### Step 5 - Any additional setup

In some cases, your package users may need to manually
install other tools in order to use your package. If that
is the case, be sure to add a section on additional setup
to your README file.

Here, briefly document (or link to documentation for) any
additional setup that is required to use your package.
This might include:

> * authentication information if it is applicable to your package.

Reviewer: ```suggestion
* authentication information, if it is applicable to your package.
```

> * additional tool installations such as GDAL.

Reviewer: ```suggestion
* additional tool installations, such as GDAL.
```

:::{note}
Many packages won't need this section in their README. In that case you can always skip this section!
:::

### Step 6 - Add a get started section

Next add a getting started section that shows how to use your package. This

> section should include a small code chunk that demonstrates importing and using

Reviewer 2: I would stress that *working* code examples are the best kind. Writing down some pseudo-code and waiting for new users to fill in the "obvious" holes will just lead to less users.

Reviewer 2: That tip looks perfect 🎉

Author: excellent. i'll resolve this now!

some of the functionality in your package.

For the pyosPackage, a short get started demo might look like this:

```python
>>> from pyospackage.add_numbers import add_num
>>> add_num(1, 2)
3

> ``````

Reviewer: ```suggestion
```
```
something wonky is going on with the formatting after this point-I think this should fix it (but could be wrong)

Reviewer 2: looks like it is too many tics, ` ``` ` to match line 121 should do it

> Or it could simply be a link to a get started tutorial that you have created. If

Reviewer: ```suggestion
Or it could simply be a link to a getting started tutorial that you have created. If
```

you don't have this yet, you can leave it empty for the time being.

This would
also be a great place to add links to any tutorials that you have created that
help users understand how to use your package for common workflows.

### Step 7 - Community section

> The community section of your README file is a place to include information for users who may want to engage with your project. This engagement will likely happen either on a platform like GitHub or GitLab.

Reviewer: ```suggestion
The community section of your README file is a place to include information for users who may want to engage with your project. This engagement will likely happen on a platform like GitHub or GitLab.
```

In the community section, you will add links to your contributing guide
and `CODE_OF_CONDUCT.md`. You will add a [`CODE_OF_CONDUCT.md` file in the next lesson](add-license-coc).

As your package grows you may also have a link to a development guide that contributors and your maintainer team will follow.

### Step 8 - Citation & License information

> Finally it is important to let users know

Reviewer: ```suggestion
Finally it is important to let users know:
```

1. how to cite your package and

> 2. what the license is.

Reviewer 2: I'm not sure I have ever seen a README.md mention the project's license, even in passing. Is this a common scientific practice?

Author: i'll post about this in our slack. the new github interface has cool little tabs when looking at the readme that make adding license information obsolete. i just looked at numpy and they do NOT have the license in their readme. I've always had it in projects i worked on but that doesn't mean it's an actual practice. let's see what folks think. 
<img width="871" alt="Screenshot 2024-02-13 at 5 04 41 PM" src="https://github.com/pyOpenSci/python-package-guide/assets/7649194/33660957-5cca-4532-b7ea-cd00f7c0d7d2">

Author: oops commented in the wrong place. fixing this locally now!

> You will create a license file for your package in this lesson.

Reviewer: I think this is a link to a future lesson? if not, text should read something like "in a future lesson"

Author: oops. that lesson is in this pr. i'll add the link!

Author: Fixed locally. great catch!

Your finished `README.md` file should look something like this:

````markdown
# pyOpenSci-package

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.8365068.svg)](https://doi.org/10.5281/zenodo.8365068)
[![pyOpenSci](https://tinyurl.com/y22nb8up)](https://github.com/pyOpenSci/software-review/issues/115)

## What packagename does

> Short description here using non technical language that describes what your package does.

Reviewer 2: ```suggestion
Short description here using non-technical language that describes what your package does.
```

## How to install

<todo - when i add more to the pyos package this can use that readme>

> To install this package... use:

Reviewer 2: ```suggestion
To install this package run:
```

`pip install packagename`

## OPTIONAL - if you have additional setup instructions add them here. if not, skip this section.

## Get started using packagename

Here add a quick code demo showing a user how to use the package after it is installed.

```
from packagename.module import xmethod

a = xmethod.dosomething(var1, var2)

> ```

Reviewer 2: I think it would feel more consistent to see the same example given above, in Step 6.

Author: good catch - fixed!!

> You can also add any links to this section to tutorials in your documentation.

Reviewer 2: ```suggestion
You can also add any to tutorials in your documentation here.
```

## Community

> Add information here about contributing to your package. Be sure to add links to your `CODE_OF_CONDUCT` file and your development guide. For now this section might be empty. You can go back and fill it in later.

Reviewer 2: ```suggestion
Add information here about contributing to your package. Be sure to add links to your `CODE_OF_CONDUCT.md` file and your development guide. For now this section might be empty. You can go back and fill it in later.
```

## How to cite packagename

citation information here
````

## <i class="fa-solid fa-hands-bubbles"></i> Wrap up

It's important to consider the information that a new user or contributor might
need when creating your `README.md` file. While there is no perfect template,
above is a set of recommendations as you are just getting started. You may find
the need for other elements to be added to this file as you further develop your
package and as a community begins to use your package.

In the [next lesson](add-license-coc), you will add a LICENSE file to
your Python package. A license file is critical as it tells users

> how they legally can (and can't use your package). It also:

Reviewer 2: ```suggestion
how they legally can (and can't) use your package. It also:
```

> 

Reviewer: it might be helpful to link to example READMEs that we think are good exemplars (perhaps from some of our accepted packages and//or community partners?).

Author: agreed! hopefully once our example package is fleshed out we will have that one. but i love the idea of using others. maybe we can get folks in slack to provide examples that they think are good and following the criteria (generally) here? the one challenge is i tried to simplify this and so the sections are not necessarily required to be in this order with this content. so it might be tough to find examples that also map to our lesson unless we create it. 

let's chat more!

````markdown
# pyOpenSci-package

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.8365068.svg)](https://doi.org/10.5281/zenodo.8365068)
[![pyOpenSci](https://tinyurl.com/y22nb8up)](https://github.com/pyOpenSci/software-review/issues/115)

## What packagename does

Short description here using non-technical language that describes what your package does.

## How to install

<todo - when i add more to the pyos package this can use that readme>
To install this package run:

`pip install packagename`

## OPTIONAL - if you have additional setup instructions add them here. if not, skip this section.

## Get started using packagename

Here add a quick code demo showing a user how to use the package after it is installed.

```python
>>> from pyospackage.add_numbers import add_num
>>> add_num(1, 2)
3

```

You can also add any to tutorials in your documentation here.

## Community

Add information here about contributing to your package. Be sure to add links to your `CODE_OF_CONDUCT.md` file and your development guide. For now this section might be empty. You can go back and fill it in later.

## How to cite packagename

citation information here

> ````

Reviewer 2: ```suggestion
```
This fence opener doesn't match any fence closer. I don't think it is needed
